### PR TITLE
First version of sampling loop.

### PR DESCRIPTION
**Pull request by Kerilk**
This adds an event loop to THAPI, allowing registering watchdogs. This will allow adding sampling watchdogs to monitor things like energy consumption.

### CODE REVIEW

**Comment by Kerilk** [comment]
@perarnau couldn't add you as a reviewer, so tagging you.

File: sampling/thapi_sampling.c [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+    struct sampling_entry **entry = NULL;
+
+    pthread_mutex_lock(&thapi_sampling_mutex);
+    clock_gettime(CLOCK_REALTIME, &now);
+    while ((entry = (struct sampling_entry **)utarray_next(thapi_sampling_events, entry)) &&
+           time_cmp(&(*entry)->next, &now) < 0) {
+      (*entry)->pfn();
+      time_add(&(*entry)->next, &(*entry)->next, &(*entry)->interval);
+      if(time_cmp(&(*entry)->next, &now) < 0)
+	      time_add(&(*entry)->next, &now, &(*entry)->interval);
```
Comment (TApplencourt) [inline sampling/thapi_sampling.c:78]: bad indentation

File: sampling/gen_sampling_custom_probes.rb [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+#include "lttng/tracepoint_gen.h"
+
+EOF
+enums = h["enums"]
+
+if enums
+  enums.each { |e|
+    LTTng.print_enum(namespace, e)
+  }
+end
```
Comment (Kerilk) [inline sampling/gen_sampling_custom_probes.rb:28]:
```suggestion

h.fetch(enums,[]).each { |e|
  LTTng.print_enum(namespace, e)
}
```

File: sampling/gen_sampling_custom_probes.rb [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+h["events"].each { |e|
```
Comment (Kerilk) [inline sampling/gen_sampling_custom_probes.rb:30]:
```suggestion
h.fetch("events",[]).each { |e|
```

File: sampling/thapi_sampling.c (hunk 1 above)
Comment (Kerilk) [inline sampling/thapi_sampling.c:78]:
```suggestion
        time_add(&(*entry)->next, &now, &(*entry)->interval);
```

File: sampling/thapi_sampling.c [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+
+void thapi_sampling_init_once() {
+  struct timespec interval;
+  utarray_new(thapi_sampling_events, &ut_ptr_icd);
+  if (!thapi_sampling_events)
+    return;
+  if (getenv("LTTNG_UST_SAMPLING_HEARTBEAT")) {
+    interval.tv_sec = 0;
+    interval.tv_nsec = 100000000;
+    thapi_register_sampling(&thapi_sampling_heartbeat, &interval);
```
[the remark this answers is not in the document]
Comment (Kerilk) [reply sampling/thapi_sampling.c:106]: They can, see below (these are only for debugging purposes, they will get removed at some point.

File: sampling/thapi_sampling.h [hunk 5]
```diff
@@ -0,0 +1,5 @@
+#include <time.h>
+
+extern int thapi_sampling_init();
+
+extern void thapi_register_sampling(void (*pfn)(void), struct timespec *interval);
```
Comment (Kerilk) [inline sampling/thapi_sampling.h:5]:
```suggestion
/*
 * TODO: Updtae API to have it take a state:
 * rename and change siganture to:
 * extern int thapi_register_sampling_callback(struct timespec *interval, int (*pfn)(void*), void* user_data);
 * so Callbakcs can take an additional state and return an status code (that could be used to stop
 * unregister the callabck)
 */
extern void thapi_register_sampling(void (*pfn)(void), struct timespec *interval);
```
Adding this here for later.

File: utils/babeltrace_energy.in [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+require 'babeltrace2'
+require 'find'
+require 'yaml'
+require 'pp'
+
+$event_lambdas = {}
+#require 'babeltrace_opencl_lib'
+#require 'babeltrace_ze_lib'
+#require 'babeltrace_cuda_lib'
+#require 'babeltrace_omp_lib'
```
Comment (Kerilk) [inline utils/babeltrace_energy.in:14]:
```suggestion
```

File: utils/babeltrace_energy.in [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+#          common_context_field = e.get_common_context_field
+#          str << " - " << common_context_field.value.collect { |k, v| "#{k}: #{v}" }.join(", ") if common_context_field
+#        end
+#        str << " - #{e.name}: "
+#        if l
+#          str << l.call(e.payload_field.value)
+#        else
+#          str << e.payload_field.to_s
+#        end
+#        puts str
```
Comment (Kerilk) [inline utils/babeltrace_energy.in:111]:
```suggestion
```

File: utils/babeltrace_energy.in [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+  $energies[key] = [energy, timestamp] 
+}
+
+consume = lambda { |iterator, _|
+  mess = iterator.next_messages
+  mess.each { |m|
+    if m.type == :BT_MESSAGE_TYPE_EVENT
+      
+      e = m.event
+      #puts e.name
```
Comment (Kerilk) [inline utils/babeltrace_energy.in:95]:
```suggestion
```

File: utils/xprof_utils.cpp [hunk 9]
```diff
@@ -20,6 +20,104 @@ thread_id_t borrow_thread_id(const bt_event *event){
     return bt_field_integer_unsigned_get_value(field);
 }
 
+
+bt_message* create_energy_message(const char* hostname, const process_id_t process_id, const thread_id_t thread_id,
+                                     const uintptr_t hDevice, const uint32_t domain, const uint64_t energy, const uint64_t ts,
+                                     bt_event_class *event_class, bt_self_message_iterator *message_iterator, bt_stream *stream, backend_t backend) {
```
Comment (Kerilk) [inline utils/xprof_utils.cpp:26]:
```suggestion
bt_message* create_power_message(const char* hostname, const process_id_t process_id, const thread_id_t thread_id,
                                 const uintptr_t hDevice, const uint32_t domain, const uint64_t energy, const uint64_t ts,
                                 bt_event_class *event_class, bt_self_message_iterator *message_iterator, bt_stream *stream, backend_t backend) {
```
Your reading power now

File: utils/xprof_utils.cpp [hunk 10]
```diff
@@ -20,6 +20,104 @@ thread_id_t borrow_thread_id(const bt_event *event){
...
+     bt_field *device_id_field = bt_field_structure_borrow_member_field_by_index(payload_field,0);
+     bt_field_integer_unsigned_set_value(device_id_field, hDevice);
+
+     // domain
+     bt_field *domain_field = bt_field_structure_borrow_member_field_by_index(payload_field,1);
+     bt_field_integer_unsigned_set_value(domain_field, domain);
+
+     // energy
+     bt_field *energy_field = bt_field_structure_borrow_member_field_by_index(payload_field,2);
+     bt_field_integer_unsigned_set_value(energy_field, energy);
```
Comment (Kerilk) [inline utils/xprof_utils.cpp:67]:
```suggestion
     // power
     bt_field *power_field = bt_field_structure_borrow_member_field_by_index(payload_field,2);
     bt_field_integer_unsigned_set_value(power_field, power);
```

File: utils/xprof_utils.hpp [hunk 11]
```diff
@@ -118,6 +118,16 @@ const char* borrow_hostname(const bt_event*);
 process_id_t borrow_process_id(const bt_event*);
 thread_id_t borrow_thread_id(const bt_event*);
 
+
+bt_message* create_energy_message(const char* hostname, const process_id_t proprocess_id, const thread_id_t thread_id,
+                                     const uintptr_t hDevice, const uint32_t domain, const uint64_t energy, const uint64_t ts,
+                                     bt_event_class *event_class, bt_self_message_iterator *message_iterator, bt_stream *stream, backend_t backend = BACKEND_UNKNOWN);
```
Comment (Kerilk) [inline utils/xprof_utils.hpp:124]:
```suggestion
bt_message* create_power_message(const char* hostname, const process_id_t proprocess_id, const thread_id_t thread_id,
                                     const uintptr_t hDevice, const uint32_t domain, const uint64_t power, const uint64_t ts,
                                     bt_event_class *event_class, bt_self_message_iterator *message_iterator, bt_stream *stream, backend_t backend = BACKEND_UNKNOWN);
```

File: xprof/interval.c.erb [hunk 12]
```diff
@@ -96,6 +96,8 @@ bt_component_class_initialize_method_status <%= namespace %>_dispatch_initialize
     dispatch->device_event_class = create_lttng_device_event_class_message(trace_class, stream_class);
     dispatch->traffic_event_class = create_lttng_traffic_event_class_message(trace_class, stream_class);
     dispatch->device_name_event_class = create_lttng_device_name_event_class_message(trace_class, stream_class);
+    dispatch->frequency_event_class = create_lttng_frequency_event_class_message(trace_class, stream_class);
+    dispatch->energy_event_class = create_lttng_energy_event_class_message(trace_class, stream_class);
```
Comment (Kerilk) [inline xprof/interval.c.erb:100]:
```suggestion
    dispatch->power_event_class = create_lttng_power_event_class_message(trace_class, stream_class);
```

File: xprof/interval.h.erb [hunk 13]
```diff
@@ -56,7 +56,8 @@ struct <%= namespace %>_dispatch {
     bt_event_class *device_event_class;
     bt_event_class *traffic_event_class;
     bt_event_class *device_name_event_class;
-
+    bt_event_class *frequency_event_class;
+    bt_event_class *energy_event_class;
```
Comment (Kerilk) [inline xprof/interval.h.erb:60]:
```suggestion
    bt_event_class *power_event_class;
```

File: xprof/interval_model.yaml [hunk 14]
```diff
@@ -53,4 +53,33 @@
...
+    :class_properties:
+      :preferred_display_base: 16
+  - :name: domain
+    :class: unsigned
+    :class_properties:
+      :preferred_display_base: 16
+  - :name: energy 
+    :class: unsigned
+    :class_properties:
+      :preferred_display_base: 16
```
Comment (Kerilk) [inline xprof/interval_model.yaml:84]:
```suggestion
- :name: lttng:power
  :payload:
  - :name: did
    :class: unsigned
    :class_properties:
      :preferred_display_base: 16
  - :name: domain
    :class: unsigned
    :class_properties:
      :preferred_display_base: 16
  - :name: power 
    :class: unsigned
```

File: xprof/interval_model.yaml [hunk 15]
```diff
@@ -53,4 +53,33 @@
...
+  - :name: domain
+    :class: unsigned
+    :class_properties:
+      :preferred_display_base: 16
+  - :name: frequency
+    :class: unsigned
+    :class_properties:
+      :preferred_display_base: 16
```
Comment (Kerilk) [inline xprof/interval_model.yaml:70]:
```suggestion
  - :name: frequency
    :class: unsigned
```
Frquency (and power) don't really benefit from being displayed in hexadecimal notation.

File: xprof/timeline.cpp [hunk 16]
```diff
@@ -5,14 +5,23 @@
...
   static std::atomic<perfetto_uuid_t> uuid{1};
   return uuid++;
 }
+bool pushIfNotExists(std::vector<uint64_t>& vec, uint64_t value) {
+    if (std::find(vec.begin(), vec.end(), value) == vec.end()) {
+        vec.push_back(value);
+        return true;
+    }
+    return false;
+}
```
Comment (Kerilk) [inline xprof/timeline.cpp:24]:
```suggestion

static bool pushIfNotExists(std::vector<uint64_t>& vec, uint64_t value) {
  if (std::find(vec.begin(), vec.end(), value) == vec.end()) {
    vec.push_back(value);
    return true;
  }
  return false;
}
```
Ideally, it should use a set: https://cplusplus.com/reference/set/set/ I think

File: xprof/timeline.cpp [hunk 17]
```diff
@@ -32,6 +41,7 @@ static void add_event_end(struct timeline_dispatch *dispatch, perfetto_uuid_t uu
   auto *track_event = packet->mutable_track_event();
   track_event->set_type(perfetto_pruned::TrackEvent::TYPE_SLICE_END);
   track_event->set_track_uuid(uuid);
+
```
Comment (Kerilk) [inline xprof/timeline.cpp:44]:
```suggestion
```

File: xprof/timeline.cpp [hunk 18]
```diff
@@ -112,6 +122,97 @@ static perfetto_uuid_t get_parent_uuid(struct timeline_dispatch *dispatch, std::
...
+    auto *packet = dispatch->trace.add_packet();
+    packet->set_trusted_packet_sequence_id(10000);
+    packet->set_timestamp(timestamp); 
+    auto *track_event = packet->mutable_track_event();
+    track_event->set_type(perfetto_pruned::TrackEvent::TYPE_COUNTER);
+    track_event->set_track_uuid(hDevice);
+    track_event->set_name("frequency");
+    track_event->set_counter_value(frequency);
+}
```
Comment (Kerilk) [inline xprof/timeline.cpp:171]:
```suggestion
static void add_event_freq(struct timeline_dispatch *dispatch, std::string hostname,  uint64_t process_id, uint64_t thread_id, uintptr_t hDevice,
                           uint32_t domain, uint64_t timestamp, uint64_t frequency) {

  if ( dev_list.empty()){
    auto *packet = dispatch->trace.add_packet(); 
    packet->set_trusted_packet_sequence_id(10000);
    packet->set_timestamp(0);
    packet->set_previous_packet_dropped(true);
    auto *track_descriptor = packet->mutable_track_descriptor();
    track_descriptor->set_uuid(1388);
    auto *process = track_descriptor->mutable_process();
    process->set_pid(1024);
    std::string name = "GPU_Frequency | " + hostname; 
    process->set_process_name(name);
  }
  // we use vector to store the unique devices, Need change to support for multiple node
  if (pushIfNotExists(dev_list, hDevice))
  {
    auto *packet = dispatch->trace.add_packet();
    packet->set_timestamp(0);
    packet->set_trusted_packet_sequence_id(10000);
    auto *track_descriptor = packet->mutable_track_descriptor();
    track_descriptor->set_uuid(hDevice);
    track_descriptor->set_parent_uuid(1388);
    std::string s="Device-" + std::to_string(hDevice);
    track_descriptor->set_name(s);
    track_descriptor->mutable_counter();
  }

  auto *packet = dispatch->trace.add_packet();
  packet->set_trusted_packet_sequence_id(10000);
  packet->set_timestamp(timestamp);
  auto *track_event = packet->mutable_track_event();
  track_event->set_type(perfetto_pruned::TrackEvent::TYPE_COUNTER);
  track_event->set_track_uuid(hDevice);
  track_event->set_name("frequency");
  track_event->set_counter_value(frequency);
}
```
Reindented

**Pull request closed without merging**